source/common/filesystem/win32/watcher_impl.cc

```diff
@@ -0,0 +1,274 @@
+  ASSERT(result.rc_ == 0);
+  result = os_sys_calls_.setsocketblocking(event_write_, false);
+  ASSERT(result.rc_ == 0);
+
+  directory_event_ = dispatcher.createFileEvent(
+      event_read_,
+      [this](uint32_t events) -> void {
+        ASSERT(events == Event::FileReadyType::Read);
+        onDirectoryEvent();
+      },
+      Event::FileTriggerType::Level, Event::FileReadyType::Read);
+
+  thread_exit_event_ = ::CreateEvent(nullptr, false, false, nullptr);
+  ASSERT(thread_exit_event_ != NULL);
+  keep_watching_ = true;
+  watch_thread_ = thread_factory_.createThread([this]() -> void { watchLoop(); });
+}
+
+WatcherImpl::~WatcherImpl() {
+  const BOOL rc = ::SetEvent(thread_exit_event_);
+  ASSERT(rc);
+
+  watch_thread_->join();
+
+  for (auto& entry : callback_map_) {
+    ::CloseHandle(entry.second->hDir_);
+    ::CloseHandle(entry.second->op_.hEvent);
+  }
+  ::CloseHandle(thread_exit_event_);
+  ::closesocket(event_read_);
+  ::closesocket(event_write_);
+}
+
+void WatcherImpl::addWatch(const std::string& path, uint32_t events, OnChangedCb cb) {
+  std::string dir_narrow(path);
+  std::string file_narrow;
+  file_system_.splitFileName(dir_narrow, file_narrow);
+  // ReadDirectoryChangesW only has a Unicode version, so we need
+  // to use wide strings here
+  const std::wstring directory = cvt_.from_bytes(dir_narrow);
+  const std::wstring file = cvt_.from_bytes(file_narrow);
+
+  const HANDLE hDir = CreateFileW(
+      directory.c_str(), GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+      nullptr, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED, NULL);
+  if (hDir == INVALID_HANDLE_VALUE) {
+    throw EnvoyException(
+        fmt::format("unable to open directory {}: {}", dir_narrow, GetLastError()));
+  }
+  std::string fii_key(sizeof(FILE_ID_INFO), '\0');
+  if (!GetFileInformationByHandleEx(hDir, FileIdInfo, &fii_key[0], sizeof(FILE_ID_INFO))) {
+    throw EnvoyException(
+        fmt::format("unable to identify directory {}: {}", dir_narrow, GetLastError()));
+  }
+
+  if (callback_map_.find(fii_key) != callback_map_.end()) {
+    CloseHandle(hDir);
+  } else {
+    callback_map_[fii_key] = std::make_unique<DirectoryWatch>();
+    callback_map_[fii_key]->hDir_ = hDir;
+    callback_map_[fii_key]->buffer_.resize(16384);
+    callback_map_[fii_key]->watcher_ = this;
+
+    // According to Microsoft docs, "the hEvent member of the OVERLAPPED structure is not used by
+    // the system, so you can use it yourself". We will use it for synchronization of the completion
+    // routines
+    HANDLE hEvent = ::CreateEvent(nullptr, false, false, nullptr);
+    if (hEvent == NULL) {
+      throw EnvoyException(fmt::format("CreateEvent failed: {}", GetLastError()));
+    }
+
+    callback_map_[fii_key]->op_.hEvent = hEvent;
+    dir_watch_complete_events_.push_back(hEvent);
+
+    // send the first ReadDirectoryChangesW request to our watch thread. This ensures that all of
```

Comment:
nit: start sentences with capital letter, end with period, proper grammar, etc. here and elsewhere.